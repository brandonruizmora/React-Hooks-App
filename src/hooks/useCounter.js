import { useState } from 'react'

export const useCounter = (initialState = 10) => {

    const [state, setState] = useState(initialState);

    const increment = function ( factor = 1 ) {
        setState( state + factor );
    }

    const decrement = function ( factor = 1 ) {
        setState( state - factor );
    }

    const reset = function () {
        setState(initialState);
    }

    return {
        state,
        increment,
        decrement,
        reset
    };

}
