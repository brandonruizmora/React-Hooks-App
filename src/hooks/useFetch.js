import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {

    const isMounted = useRef(true);

    const [state, setState] = useState({ data: null, loading: true, error: null });

    useEffect(() => {

        return () => {
            isMounted.current = false;
        }

    }, []);

    useEffect(() => {

        setState({
            loading: true,
            error: null,
            data: null
        });

        fetch(url)
            .then(res => res.json())
            .then(data => {
                if (isMounted) {
                    setTimeout(() => {
                        setState({
                            loading: false,
                            error: null,
                            data
                        });
                    }, 3000);
                } else {
                    console.log('setState no se llamo');
                }
            });

    }, [url]);

    return state;

}
