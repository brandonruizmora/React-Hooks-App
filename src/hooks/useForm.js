import { useState } from "react"

export const useForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState)

    const handleInputChange = function ({ target }) {
        setValues({
            ...values,
            [target.name]: target.value
        });
    }

    return [ values, handleInputChange ]
    
}
