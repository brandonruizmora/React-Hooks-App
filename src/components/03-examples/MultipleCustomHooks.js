import React from 'react';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const { counter, increment, decrement } = useCounter(1);
    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
    const { author, quote } = !!data && data[0];

    console.log(loading)

    return (
        <div className="m-5">
            <h1>BreakingBad Quotes</h1>
            <hr />
            {loading ?
                (
                    <div className="alert alert-primary" role="alert">
                        Loading...
                    </div>
                )
                :
                (
                    <figure className="text-end">
                        <blockquote className="blockquote">
                            <p>{ quote }</p>
                        </blockquote>
                        <figcaption className="blockquote-footer">
                            { author }
                        </figcaption>
                    </figure>
                )
            }
            <button className="btn btn-primary mx-1" onClick={ () => decrement(1) }>
                Anterior
            </button>
            <button className="btn btn-success mx-1" onClick={ () => increment(1) }>
                Siguiente
            </button>
        </div>
    )
}
