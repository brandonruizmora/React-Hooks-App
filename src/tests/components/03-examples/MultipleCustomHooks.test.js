import '../../../setupTest';
import { shallow } from "enzyme";
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { useFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas en <MultipleCustomHooks />', () => {

    test('Debe mostrarse correctamente', () => {
        useCounter.mockReturnValue({
            counter: 1,
            increment: () => {}
        });
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });
        const wrapper = shallow( <MultipleCustomHooks /> );
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrar la informacion', () => {
        useCounter.mockReturnValue({
            counter: 1,
            increment: () => {}
        });
        useFetch.mockReturnValue({
            data: [{
                author: 'Brandon',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow( <MultipleCustomHooks /> );

        expect( wrapper.find('.alert').exists() ).toBe( false );
        expect( wrapper.find('p').text() ).toBe( 'Hola Mundo' );
        // console.log( wrapper.find('button').at(1).html());
    });
    
});
