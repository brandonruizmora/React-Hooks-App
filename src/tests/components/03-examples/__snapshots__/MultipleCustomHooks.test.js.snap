// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe mostrarse correctamente 1`] = `
<div
  className="m-5"
>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-primary"
    role="alert"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary mx-1"
    onClick={[Function]}
  >
    Anterior
  </button>
  <button
    className="btn btn-success mx-1"
    onClick={[Function]}
  >
    Siguiente
  </button>
</div>
`;
